diff-containers/test/Test/Data/Map/Diff/Strict.hs

```diff
@@ -43,123 +46,219 @@ import           Test.Util
+     (Ord k, Eq v)
+  => NormalDiff k v
+  -> NormalDiff k v
+  -> Property
+prop_summingDiffsPreservesNormality (NormalDiff d1) (NormalDiff d2) =
+  isNormal d1 && isNormal d2 ==> isNormal (d1 <> d2)
+
+-- | Test the invariant that summing @'DiffHistory'@s preserves normality.
+prop_summingDiffHistoriesPreservesNormality ::
+     Eq v
+  => NormalDiffHistory v
+  -> NormalDiffHistory v
+  -> Property
+prop_summingDiffHistoriesPreservesNormality (NormalDiffHistory h1) (NormalDiffHistory h2) =
+  isNormalDiffHistory h1 && isNormalDiffHistory h2 ==> isNormalDiffHistory (h1 <> h2)
+
+-- | Test the invariant that inverting @'Diff'@s preserves normality.
+prop_invertingDiffsPreservesNormality ::
+     (Ord k, Eq v)
+  => NormalDiff k v
+  -> Property
+prop_invertingDiffsPreservesNormality (NormalDiff d) =
+  isNormal d ==> isNormal (invert d)
+
+-- | Test the invariant that inverting @'DiffHistory'@s preserves normality.
+prop_invertingDiffHistoriesPreservesNormality ::
+     Eq v
+  => NormalDiffHistory v
+  -> Property
+prop_invertingDiffHistoriesPreservesNormality (NormalDiffHistory h) =
+  isNormalDiffHistory h ==> isNormalDiffHistory (invert h)
+
+-- | Test the invariant that the identity @'Diff'@ element is positive.
+prop_emptyDiffIsPositive :: Property
+prop_emptyDiffIsPositive = once $ isPositive (mempty :: Diff Int Int)
+
+-- | Test the invariant that the identity @'DiffHistory'@ element is positive.
+prop_emptyDiffHistoryIsPositive :: Property
+prop_emptyDiffHistoryIsPositive = once $ isPositiveDiffHistory (mempty :: DiffHistory Int)
+
+-- | Test the invariant that the identity @'Diff'@ element is normal.
+prop_emptyDiffIsNormal :: Property
+prop_emptyDiffIsNormal = once $ isNormal (mempty :: Diff Int Int)
+
+-- | Test the invariant that the identity @'DiffHistory'@ element is normal.
+prop_emptyDiffHistoryIsNormal:: Property
+prop_emptyDiffHistoryIsNormal = once $ isNormalDiffHistory (mempty :: DiffHistory Int)
+
+-- | Applying a @'Diff'@ computed from a source and target @'Map'@ should
+-- produce the target @'Map'@.
 prop_diffThenApply ::
      (Show k, Show v, Ord k, Eq v)
   => Map k v
   -> Map k v
   -> Property
 prop_diffThenApply m1 m2 = applyDiff m1 (diff m1 m2) === Right m2
 
+-- | Applying an empty @'Diff'@ is the identity function.
 prop_applyMempty ::
      (Show k, Show v, Ord k, Eq v)
   => Map k v
   -> Property
 prop_applyMempty m = applyDiff m mempty === Right m
 
+-- | Applying a sum of normal and positive @'Diff'@s is equivalent to applying
+-- each (normal and positive) @'Diff'@ separately (in order).
 prop_applyAllAndApplySum ::
      (Show k, Show v, Ord k, Eq v)
   => Map k v
-  -> [Diff k v]
+  -> [NPDiff k v]
   -> Property
-prop_applyAllAndApplySum m ds =
-  all isPositive ds ==> foldlM applyDiff m ds === applyDiff m (mconcat ds)
+prop_applyAllAndApplySum m (fmap getNPDiff -> ds) =
+  all isPositive ds && all isNormal ds ==> foldlM applyDiff m ds === applyDiff m (mconcat ds)
```

Comment:
Maybe it's worth tagging how many cases are we filtering here (due to the `==>` operator).

Reply:
And the same for other uses of `==>` below.

Reply:
As these are "guaranteed" by the generator, I would actually either test the generators (the thing to do :tm:) or change this to an `assert` that will crash if this is not true.

Reply:
> Maybe it's worth tagging how many cases are we filtering here (due to the `==>` operator).

QuickCheck will already report how many tests are discarded, is that 

> As these are "guaranteed" by the generator, I would actually either test the generators (the thing to do :tm:) or change this to an `assert` that will crash if this is not true.

I added the implications to see whether the generated values were actually normal/positive, so that if the tests are discarded, I could see that my generators are wrong. Testing the generators and adding assertions is probably better :P